Actor builder and updates

BrandonArp opened the pull request:
Just a preview.  I need to add the GraphiteSink back in with Akka actors.  I might also pull out the updates from the ActorBuilder stuff.

BrandonArp commented:
The GraphiteSink was updated and the updated dependencies are in a separate PR. Please take a look at #52 before this.

BrandonArp commented:
Rebased off the updated dependencies.  There are a few minor changes in here that are unrelated: minor refactor of emitter creation, changing to Java-based futures in Status actor.

vjkoskela commented on src/main/java/com/arpnetworking/akka/NonJoiningClusterJoiner.java:
```diff
@@ -0,0 +1,83 @@
+ *
+ * @author Brandon Arp (brandon dot arp at inscopemetrics dot com)
+ */
+public class NonJoiningClusterJoiner extends UntypedActor {
```
Final?

BrandonArp replied:
Oh you and your final classes.

vjkoskela commented on src/main/java/com/arpnetworking/akka/NonJoiningClusterJoiner.java:
```diff
@@ -0,0 +1,83 @@
+     * @param builder Builder to create the Props from
+     * @return a new {@link Props}
+     */
+    private static Props props(final Builder builder) {
```
Builder is unused and the method is private

BrandonArp replied:
The builder references the static method as the constructor for the Props.  It looks similar to our other actors (except for the private part).  I could make the props method public, but I'm not sure what that would really achieve.

vjkoskela replied:
I guess my confusion here is that the builder argument is unused and since it's static it can't be overwritten. This is also not the standard builder pattern so it's not invoked reflectively. Still a little confused here.

BrandonArp replied:
The Builder still references this method just like the constructor in the other Builders.  The base builder takes a Function<Builder, Props>.  Hence the need to have that signature.

vjkoskela commented on src/main/java/com/arpnetworking/akka/NonJoiningClusterJoiner.java:
```diff
@@ -0,0 +1,83 @@
+     * Public constructor.
+     */
+    public NonJoiningClusterJoiner() {
+        LOGGER.info()
```
Shouldn't this go in the actor startup routine, not the ctor?

BrandonArp replied:
Eh, I could put it on the onStart.  I've tended toward the constuctor, though.

vjkoskela replied:
Truth in advertising. It doesn't "start" on construction.

vjkoskela commented on src/main/java/com/arpnetworking/akka/NonJoiningClusterJoiner.java:
```diff
@@ -0,0 +1,83 @@
+                .log();
+    }
+
+    @Override
```
Inherit doc

BrandonArp replied:
Fixed.

vjkoskela commented on src/main/java/com/arpnetworking/akka/NonJoiningClusterJoiner.java:
```diff
@@ -0,0 +1,83 @@
+            super(NonJoiningClusterJoiner::props);
+        }
+
+        @Override
```
Same.

BrandonArp replied:
Fixed.

vjkoskela commented on src/main/java/com/arpnetworking/clusteraggregator/Status.java:
```diff
@@ -108,80 +105,69 @@ public void onReceive(final Object message) throws Exception {
+                    .thenApply(CAST_MAPPER);
+            stateFuture.whenComplete(
+                    (statusResponse, throwable) -> {
+                        final boolean healthy = _cluster.readView().self().status() == MemberStatus.up() && !_quarantined;
```
Objects.equal instead of ==

BrandonArp replied:
Fixed

vjkoskela commented on src/main/java/com/arpnetworking/configuration/jackson/akka/ActorBuilderDeserializer.java:
```diff
@@ -0,0 +1,69 @@
+        try {
+            final Class<?> clazz = Class.forName(type);
+            final Class<? extends Builder<? extends Props>> builder = getBuilderForClass(clazz);
+            final Builder<? extends Props> value = _mapper.readValue(treeNode.toString(), builder);
```
You should be able to avoid the toString here

BrandonArp replied:
I think I found a suitable alternative.  Docs indicate that it should be more efficient.

vjkoskela commented on src/main/java/com/arpnetworking/configuration/jackson/akka/ActorBuilderDeserializer.java:
```diff
@@ -0,0 +1,69 @@
+    @Override
+    public Props deserialize(final JsonParser p, final DeserializationContext ctxt) throws IOException {
+        final TreeNode treeNode = p.readValueAsTree();
+        final String type = ((TextNode) treeNode.get("type")).textValue();
```
This seems like a bit of reinvention. Type mapping is already supported by Jackson, I know we haven't used it, but I think so far we've avoided this level of generality. For example we have a getClass helper from config which we then use to deserialize but we never fetch the builder by hand. Could we just do something like that here?

BrandonArp replied:
I tried, but I couldn't pull it off.  This was the best I was able to do.  Everything else got.... interestingly broken.  It had to do with the return type all being Props, but the deserializer not being a subclass of Props and the fact that each Props will have a different constructor.  I don't think Jackson was setup to do that.

vjkoskela replied:
Hrm. Ok.

The use case seems to me be a combination of 1) Polymorphism; different builder types as determined by a field; 2) The use of a builder. In the first we've either done explicit type request via a TypeReference or annotation based binding via a type field. 

Props is final so you can't wrap it or extend it in order to annotate it.

http://doc.akka.io/japi/akka/2.4.0/akka/actor/Props.html

The problem seems to be that we're storing any Props here for the _clusterJoinerActor which is not type-complete (I'm guess not just any actor can be a "joiner actor").

For example, if the Props were wrapped with NonJoiningClusterJoiner and we used that type here (or more correctly an interface of that type) then we would be just to a manageable polymorphic deserialization problem (e.g. annotate the subclasses of that interface). Using Props directly corners us in many ways, and doesn't help describe what is actually required here.

Happy to discuss.

vjkoskela commented on src/main/java/com/arpnetworking/clusteraggregator/configuration/ClusterAggregatorConfiguration.java:
```diff
@@ -429,6 +455,8 @@ public Builder setDatabaseConfigurations(final Map<String, DatabaseConfiguration
         @NotNull
         private String _clusterHostSuffix = "";
+        @NotNull
+        private Props _clusterJoinActor = new NonJoiningClusterJoiner.Builder().build();
```
This isn't actually an actor. Do we want it to be? In the past we've done that through Guice, but as this is configuration I am inclined to say it's Props, but the name should probably reflect that.

BrandonArp replied:
At the end of the day it becomes an actor.  I don't want the people configuring the service to have to know that they're actually building a Props that will get created into an actor.  As far as they know, they want to configure an actor.

BrandonArp commented:
I've addressed most of the concerns.  The implementation doesn't feel quite right still, but the semantics of the config file match what I want.  It feels like there is still something to do for the declared type on the config classes, or perhaps in the way it polymorphically deserialize (or, rather, the way it doesn't use the built-in Jackson polymorphic deserialization).
